File: android/src/main/kotlin/com/flutter_timezone_observer/FlutterTimezoneObserverPlugin.kt
package com.flutter_timezone_observer

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import java.time.ZoneId

/**
 * [FlutterPlugin] for the flutter_timezone_observer package.
 *
 * This class handles the native Android implementation for fetching the
 * current timezone and listening for timezone changes.
 */
class FlutterTimezoneObserverPlugin : FlutterPlugin, EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel

    private var applicationContext: Context? = null
    private var timezoneReceiver: BroadcastReceiver? = null

    private val METHOD_CHANNEL_NAME = "flutter_timezone_observer/methods"
    private val EVENT_CHANNEL_NAME = "flutter_timezone_observer/events"

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = flutterPluginBinding.applicationContext

        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, METHOD_CHANNEL_NAME)
        methodChannel.setMethodCallHandler { call, result ->
            if (call.method == "getLocalTimezone") {
                try {
                    val timezone = ZoneId.systemDefault().id
                    result.success(timezone)
                } catch (e: Exception) {
                    result.error("ERROR", "Failed to get timezone", e.toString())
                }
            } else {
                result.notImplemented()
            }
        }

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, EVENT_CHANNEL_NAME)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        applicationContext = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        timezoneReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                if (intent?.action == Intent.ACTION_TIMEZONE_CHANGED) {
                    val timezone = ZoneId.systemDefault().id
                    // Send the new timezone to Flutter
                    events?.success(timezone)
                }
            }
        }
        applicationContext?.registerReceiver(
            timezoneReceiver,
            IntentFilter(Intent.ACTION_TIMEZONE_CHANGED)
        )
    }

    override fun onCancel(arguments: Any?) {
        if (timezoneReceiver != null) {
            applicationContext?.unregisterReceiver(timezoneReceiver)
            timezoneReceiver = null
        }
    }
}

File: ios/Classes/FlutterTimezoneObserverPlugin.swift
import Flutter
import UIKit

/**
 * [FlutterPlugin] for the flutter_timezone_observer package.
 *
 * This class handles the native iOS implementation for fetching the
 * current timezone and listening for timezone changes.
 */
public class FlutterTimezoneObserverPlugin: NSObject, FlutterPlugin {

    private var timezoneStreamHandler: TimezoneStreamHandler?

    private let METHOD_CHANNEL_NAME = "flutter_timezone_observer/methods"
    private let EVENT_CHANNEL_NAME = "flutter_timezone_observer/events"

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FlutterTimezoneObserverPlugin()

        let methodChannel = FlutterMethodChannel(name: instance.METHOD_CHANNEL_NAME,
                                                 binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(name: instance.EVENT_CHANNEL_NAME,
                                               binaryMessenger: registrar.messenger())

        instance.timezoneStreamHandler = TimezoneStreamHandler()
        eventChannel.setStreamHandler(instance.timezoneStreamHandler)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "getLocalTimezone" {
            let timezone = TimeZone.current.identifier
            result(timezone)
        } else {
            result(FlutterMethodNotImplemented)
        }
    }
}

/**
 * A dedicated [FlutterStreamHandler] for timezone change events.
 *
 * This class manages the lifecycle of the event stream, listening for
 * [NSSystemTimeZoneDidChange] notifications.
 */
class TimezoneStreamHandler: NSObject, FlutterStreamHandler {

    private var eventSink: FlutterEventSink?

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(onTimezoneChanged),
            name: NSNotification.Name.NSSystemTimeZoneDidChange,
            object: nil
        )
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        NotificationCenter.default.removeObserver(self)
        self.eventSink = nil
        return nil
    }

    @objc func onTimezoneChanged(notification: NSNotification) {
        let timezone = TimeZone.current.identifier
        self.eventSink?(timezone)
    }
}